synapse/handlers/presence.py

```diff
@@ -445,20 +445,18 @@ def send_user_sync(self, user_id: str, is_syncing: bool, last_sync_ms: int) -> N
             )
 
     def mark_as_coming_online(self, user_id: str) -> None:
-        """A user has started syncing. Send a UserSync to the presence writer,
-        unless they had recently stopped syncing.
-        """
-        going_offline = self.users_going_offline.pop(user_id, None)
-        if not going_offline:
-            # Safe to skip because we haven't yet told the presence writer they
-            # were offline
-            self.send_user_sync(user_id, True, self.clock.time_msec())
+        """A user has started syncing. Send a UserSync to the presence writer."""
+        # Need to remove the user_id from users_going_offline or the looping call will
+        # tell the presence writer that the user went offline.
+        self.users_going_offline.pop(user_id, None)
+        self.send_user_sync(user_id, True, self.clock.time_msec())
```

Comment:
I'm not quite following why we remove the if-statement here. Can you explain a bit more what this is doing?

Reply:
I mentioned that above in Number 2 under Three-fold fix.
Some context to keep in mind before I begin the long version:
* This is only applies on a worker model of synapse where there is either a sync worker or a presence writer(at the very least). This does not apply on a monolith deployment.
* syncs can happen sometimes in rapid or sometime in slow succession, depending on how many other users the homeserver is aware exist + any room updates coming along
* There is a looping call every 10 seconds, that will mark a user as 'offline' if they are in `users_going_offline` and at least 10 seconds has passed. These timers are not synchronized in any way and a theoretical 19 seconds could occur. The looping call uses the companion to `send_user_sync()` called `send_stop_syncing()` which makes an explicit call to the presence writer to mark them 'offline'.
* This is tied to the context manager that wraps a call to the sync handler [here](https://github.com/matrix-org/synapse/blob/ca5d5de79b203f41e0fafa5ef1e7c8d28f63049d/synapse/rest/client/sync.py#L206-L219)

Now

1. At the `_end()` of the sync context manager, it used to add `user_id` as a key to `users_going_offline` with a value of the current time in milliseconds.
2. On the next sync, `user_id` is `pop()`ped off of `users_going_offline` and checked for `None`(or more specifically, would check that there was `not` a value), and it would only then `send_user_sync()`. This only occurs when it is the very first sync of a recent-ish session otherwise it will `not` be `None` and `send_user_sync()` would not be called.
3. So, we are mid sync and the presence writer has not been informed we are still around. Finish the context manager, and repeat however many times it takes to get to about 10 seconds(specifically the next time the looping call comes around to call `send_stop_syncing()` and mark them offline).
4. 10 seconds has passed, the context manager is still doing it's thing and knows the user is still online(obviously, they are still mid-sync after all). Over on the presence writer however, no indication has happened in 10 seconds that the user is still online: no further `USER_SYNC` has come over TCP replication, and remember that `set_state()` no longer bumps `last_active_ts` on a regular basis anymore. The machinery kicks into gear to mark the user as 'offline', which then gets sent over replication back to the sync handler side.
5. At this point, the presence writer thinks the user is 'offline' and the sync handler(well, it's associated `WorkerPresenceHandler`, actually) thinks the user is 'online'. The call from the presence writer will overwrite what the context manager thinks at this point. Still no update to the `last_active_ts`, by the way.
6. On the next sync, the context manager saw they were 'offline' last and puts the user back to 'online', beginning the cycle again. You can see this in metrics under the spoiler arrow with the words "One without Two resulted in this bad dream(every 35 seconds)".

Now, you are probably thinking, "Well, let's just remove the `not` in the conditional to check for an actual time". I considered that, but then the presence writer doesn't get it's first notification that the user is syncing and that felt like undefined behavior. Honestly, it might work out(and is a place I'm planning on revisiting in the future as a bit of refactor, there are more things to touch here). I will test this if you wish, my instinct says as long as the second sync of a session is within 10 seconds then everything will be fine. 

And, to be fair, fully fixing everything possible wasn't the scope of this PR. I wanted 'unavailable' to work as advertised. But this was affecting the behavior in a negative way and wasn't going to be allowed to get away, a traditional 'bug' that was fully exposed by `last_active_ts` not being updated on every single sync any longer.